src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentSelector.cs

```diff
@@ -103,5 +86,56 @@ private class ViewComponentCandidate
 
             public Type Type { get; set; }
         }
+
+        private class ViewComponentCandidateCache
+        {
+            private readonly ILookup<string, ViewComponentCandidate> _lookupByShortName;
+            private readonly ILookup<string, ViewComponentCandidate> _lookupByFullName;
+
+            public ViewComponentCandidateCache(ViewComponentCandidate[] candidates)
+            {
+                _lookupByShortName = candidates.ToLookup(c => c.ShortName, c => c);
+                _lookupByFullName = candidates.ToLookup(c => c.FullName, c => c);
+            }
+
+            public Type SelectByShortName(string name)
+            {
+                return Select(_lookupByShortName, name);
+            }
+
+            public Type SelectByFullName(string name)
+            {
+                return Select(_lookupByFullName, name);
+            }
+
+            private static Type Select(ILookup<string, ViewComponentCandidate> candidates, string name)
+            {
+                var matches = candidates[name];
+
+                var count = matches.Count();
+                if (count == 0)
+                {
+                    return null;
+                }
+                else if (count == 1)
```

Comment:
nit: `else if` -> `if` with a blank line above and `else` -> nada.  bit easier to read without both early exits and chained conditions.